source/extensions/quic_listeners/quiche/envoy_quic_stream.h

```diff
@@ -47,7 +47,26 @@ class EnvoyQuicStream : public Http::StreamEncoder,
       // Avoid calling this while decoding data because transient disabling and
       // enabling reading may trigger another decoding data inside the
       // callstack which messes up stream state.
-      switchStreamBlockState(disable);
+      if (disable) {
+        // Block QUIC stream right away. And if there are queued switching
+        // state callback, update the desired state as well.
+        switchStreamBlockState(true);
+        if (unblock_posted_) {
+          should_block_ = true;
+        }
+      } else {
+        should_block_ = false;
+        if (!unblock_posted_) {
+          // If this is the first time unblocking stream is desired, post a
+          // callback to do it in next loop. This is because unblocking QUIC
+          // stream can lead to immediate upstream encoding.
+          unblock_posted_ = true;
+          connection()->dispatcher().post([this] {
```

Comment:
Just checking, if something else happens in this event loop to change should_block, switchStreamBlockState will still be called.  Do we handle that correctly?